### ISSUE TITLE
Remove before callbacks

### ISSUE THREAD

**Issue by edwardloveall** [luckyframework/lucky_record#261]
Currently, when forms are saved methods are called in this order:

* **prepare**
* _validation_
* **before_save**
* **before_create** / **before_update**
* _triggering method eg. save, create, or update_
* **after_save**
* **after_create** / **after_update**

**Bold** methods are methods that users can hook into.

There are currently no use cases we can think of for `before_*` callbacks because that method is called _after_ the form is known to be valid.

The thought is, what if we just get rid of them, and add them back if there someone can present a solid use case for them.

## Alternatives

* Remove _prepare_ and move the `before_*` callbacks to they happen before validation
* Remove callbacks and create blocks inside the prepare method to distinguish between create/update steps:
```cr
def prepare
  # validations
  on_create do
    # set token
  end
end
```
* Remove callbacks and have separate `prepare_for_create/update` methods

**Comment by edwardloveall** [comment]
Here's the crappiest proof of concept for the 2nd option:

```cr
class BaseForm
  def on_create
    yield
  end
  
  def save
    prepare
  end
end

class UserForm < BaseForm
  def prepare
    on_create { save_token }
  end
  
  def save_token
    puts "Token!"
  end
end

UserForm.new.save
```

**Comment by paulcsmith** [comment]
So I was thinking about this a bit more and I think it would be cool to remove `prepare` because it is a bit of a special method. The callbacks work differently than prepare and this would be just one more different way of using prepare and callbacks. So I was thinking something like this would be used instead of prepare:

```crystal
class UserForm < User::BaseForm
  before_save :validate
  before_create :set_username
  
  def validate
    # Run some validations
  end

  def set_username
    username.value = email.value.try(&.split("@").first)
  end
end
```

I think this also makes it clearer how it works because it uses the same pattern as other callbacks. What does everyone think?

**Comment by edwardloveall** [comment]
Yeah! I like that. Prepare never felt right somehow to me. It seems like it was mostly a `before_validation` step?

Also, what do you think about the whole remove all callbacks idea I had above? Looking at it again it looks kind of slick, but I'd love to figure out a way to remove as many special case methods as possible.

Maybe you have to make your own method to save a form and call our special method whenever you want? Maybe like:

```crystal
class UserForm < User::BaseForm
  def save
    do_some_pre_validation
    do_some_other_stuff
    validate_confirmation_of # Lucky-provided method
    save_to_database         # Lucky-provided method
    something_afterwards_if_you_need_it_i_dunno
  end
end
```

This way there's _very_ little magic, it's easy to extend, it's easy to read, and it's easy to mix custom logic and pre-built methods. I'm sure I'm missing something here. Point it out!

**Comment by paulcsmith** [comment]
I think removing the magic would be freakin awesome! `prepare` was kind of a `before_validation` and `before_save` wrapped into one. The most confusing part was that it kind go blended a few things into one. It was nice not to have to type `before_save :validate`, but I think that extra bit of typing makes it more clear.

But back to removing those methods, I think there are a few things that might make this tricky.

* Adding callbacks from external libraries might be tedious. Like having to call an `audit_record` in every method. Maybe this is a good thing to be more explicit though.
* There needs to be some way to only run those things in certain circumstances. So there would still need to be something like `before_create { }` or `after_update {}` etc. that you can use.
* Might be easy to accidentally run things that should or should not be run (I'll write an example near the end)
* 

So maybe your idea would work. Something like this maybe?

```crystal
class UserForm < User::BaseForm
  def call
     validate do
       validate_confirmation/requires/etc.
     end
     after_commit { |user| WelcomeEmail.new(user).deliver_now }
  end
end
```

You could extract methods to clean it up a bit:

```crystal
class UserForm < User::BaseForm
  def call
     run_validations
     send_welcome_email
  end

  def run_validations
    validate { validate_confirmation/requires/etc }
    # Maybe something like `validate_create/update` could be added as well?
    validate_create { validate_something }
  end

  def send_welcome_email
    after_commit { |user| WelcomeEmail.new(user).deliver_now }
  end
end
```

I don't love that you need to extract another method to run validations, but maybe that is not too bad to have to do.

Here's where I think it might be confusing/error-prone:

```crystal
class UserForm < User::BaseForm
  def call
    send_welcome_email
  end
end
```

It isn't clear under what circumstances that method should be run because the author (me :P) forgot to wrap it in `validate {}` or `before/after_create/update {}`.

There would also be a way to run the default validations (like the validate_required that is automatically added). You could call `super` or `previous_def` but that would be really easy to forget :S. With the callback macros they are inherited so it is super easy to do and you can't accidentally forget it. I'm having a tough time deciding between these. My gut is saying to go with the callback macros that are a bit more magical, but also make it really clear what is going on and when. But I also like the idea of making it all just regular methods with no macros at all 🤔

**Comment by paulcsmith** [comment]
Also, after looking at those examples, I get the same vibes I get from `prepare` where a bunch of code is called but you only want it to run in certain circumstances, so it is hard to tel what is *actually* going to run. I kind of like that the callback macros make it super clear when something will be run. What do you think?

**Comment by edwardloveall** [comment]
I see what you're saying. That magic of the callbacks help you not shoot yourself in the foot... but then it's magical.

I think I prefer the named callbacks. The are well named, and a known pattern. And they aren't so crazy magical. This brings me essentially back to my idea [here](https://github.com/luckyframework/lucky_record/issues/261#issuecomment-402551981).

I'm gonna sleep on it. Maybe there's a way to use types instead of methods. Maybe we don't need macros. I just tried to modify my original idea above and failed but it's late so maybe now isn't the best time.

Anyway, to sum up, the best way to shoot yourself in the foot is with ✨ magic ✨
